### PR TITLE
Jwilaby/v4 unit test infrastructure

### PR DESCRIPTION
**Pull request by justinwilaby**
This PR provides unit testing infrastructure via jest and fixes linting issues.

### CODE REVIEW

**Comment by emadelwany** [comment]
Great change @justinwilaby! So good to see unit test infra, coverall integration, CI, and so many linting errors fixed in one PR. 

Looks like there's a simple build error though...

File: packages/app/client/src/data/http/luisApi.ts [hunk 1]
```diff
@@ -25,32 +25,33 @@
 // THE SOFTWARE IS PROVIDED ""AS IS"", WITHOUT WARRANTY OF ANY KIND,
 // EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 // MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
-// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// NONINFRINGEM,ENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
```
Comment (tonyanziano) [inline packages/app/client/src/data/http/luisApi.ts]: rogue comma here

File: packages/app/client/src/data/sagas/azureBotServiceSagas.ts [hunk 2]
```diff
@@ -74,14 +86,16 @@ function* openAzureBotServiceContextMenu(action: AzureBotServiceAction<AzureBotS
...
   const { tenantId, subscriptionId, resourceGroup, id } = action.payload.azureBotService;
-  const link = `https://ms.portal.azure.com/#@${tenantId}/resource/subscriptions/${subscriptionId}/resourceGroups/${resourceGroup}/providers/Microsoft.BotService/botServices/${id}/channels`;
-  yield CommandService.remoteCall('electron:openExternal', link);
+  const thankYouTsLint = `https://ms.portal.azure.com/#@${tenantId}/resource/subscriptions/${subscriptionId}`;
+  const link = `${thankYouTsLint}/resourceGroups/${resourceGroup}/providers/Microsoft.BotService/botServices/${id}`;
+  yield CommandService.remoteCall('electron:openExternal', link + '/channels');
 }
 
 function* removeAzureBotServiceFromActiveBot(azureBotService: IAzureBotService): IterableIterator<any> {
+  // TODO - localization
```
Comment (tonyanziano) [inline packages/app/client/src/data/sagas/azureBotServiceSagas.ts]: Link this TODO to a tracking issue please

File: packages/app/client/src/data/sagas/luisSagas.ts [hunk 3]
```diff
@@ -108,9 +117,10 @@ function* openLuisContextMenu(action: LuisServiceAction<LuisServicePayload>): It
 }
 
 function* removeLuisServiceFromActiveBot(luisService: ILuisService): IterableIterator<any> {
+  // TODO - localization
```
Comment (tonyanziano) [inline packages/app/client/src/data/sagas/luisSagas.ts]: same as above

File: packages/app/client/src/ui/dialogs/botCreationDialog.tsx [hunk 4]
```diff
@@ -226,81 +270,33 @@ export class BotCreationDialog extends React.Component<{}, BotCreationDialogStat
     ActiveBotHelper.confirmAndCreateBot(bot, secret)
       .then(() => DialogService.hideDialog())
       .catch(err => console.error('Error during confirm and create bot: ', err));
-  };
+  }
 
   private showBotSaveDialog = async (): Promise<any> => {
     // get a safe bot file name
     const botFileName = await CommandService.remoteCall('file:sanitize-string', this.state.bot.name);
+    // TODO - Localization
```
Comment (tonyanziano) [inline packages/app/client/src/ui/dialogs/botCreationDialog.tsx]: issue number

File: packages/app/client/src/ui/editor/botSettingsEditor.tsx [hunk 5]
```diff
@@ -190,56 +215,41 @@ class BotSettingsEditorComponent extends React.Component<BotSettingsEditorProps,
...
+  }
 
-  private setDirtyFlag(dirty) {
+  private setDirtyFlag(dirty: boolean) {
     store.dispatch(EditorActions.setDirtyFlag(this.props.documentId, dirty));
   }
 
   private showBotSaveDialog = async (): Promise<any> => {
     // get a safe bot file name
+    // TODO - localization
```
Comment (tonyanziano) [inline packages/app/client/src/ui/editor/botSettingsEditor.tsx]: issue number

File: packages/app/client/src/ui/editor/emulator/parts/chat.tsx [hunk 6]
```diff
@@ -157,17 +157,17 @@ class Chat extends React.Component<Props> {
 
   render() {
     const endpoint = this.getEndpoint();
-
+    // TODO - localization
```
Comment (tonyanziano) [inline packages/app/client/src/ui/editor/emulator/parts/chat.tsx]: issue number

File: packages/app/client/src/ui/helpers/activeBotHelper.ts [hunk 7]
```diff
@@ -66,12 +66,13 @@ export const ActiveBotHelper = new class {
 
   confirmCloseBot(): Promise<any> {
     const hasTabs = hasNonGlobalTabs();
+    // TODO - localization
```
Comment (tonyanziano) [inline packages/app/client/src/ui/helpers/activeBotHelper.ts]: issue number

File: packages/app/shared/package.json [hunk 8]
```diff
@@ -6,20 +6,21 @@
   "main": "built/index.js",
   "types": "built/index.d.ts",
   "scripts": {
-    "build": "npm run clean && tsc --pretty && tslint --project tsconfig.json --force",
+    "build": "tslint --project tsconfig.json && npm run clean && tsc --pretty",
```
Comment (tonyanziano) [inline packages/app/shared/package.json]: could be simplified to

``npm run lint && npm run clean && tsc --pretty"`

File: packages/emulator/core/src/types/log/item.ts [hunk 9]
```diff
@@ -70,20 +70,20 @@ export type ILogItem = {
...
 } | {
-  type: "exception",
+  type: 'exception',
   payload: {
     err: any  // Shape of `Error`, but enumerable
   }
-}
+};
 
 export default ILogItem;
```
Comment (tonyanziano) [inline packages/emulator/core/src/types/log/item.ts]: missing newline

File: packages/sdk/client/package.json [hunk 10]
```diff
@@ -5,20 +5,21 @@
   "main": "built/index.js",
   "types": "built/index.d.ts",
   "scripts": {
-    "build": "npm run clean && tsc && tslint --project tsconfig.json --force",
+    "build": "tslint --project tsconfig.json && npm run clean && tsc",
```
Comment (tonyanziano) [inline packages/sdk/client/package.json]: could be simplified to

`"npm run lint && npm run clean && tsc"`

File: packages/sdk/main/package.json [hunk 11]
```diff
@@ -5,9 +5,10 @@
   "main": "built/index.js",
   "types": "built/index.d.ts",
   "scripts": {
-    "build": "npm run clean && tsc && tslint --project tsconfig.json --force",
+    "build": "tslint --project tsconfig.json && npm run clean && tsc",
```
Comment (tonyanziano) [inline packages/sdk/main/package.json]: could be simplified to

`"npm run lint && npm run clean && tsc"`

File: packages/sdk/shared/package.json [hunk 12]
```diff
@@ -5,20 +5,21 @@
   "main": "built/index.js",
   "types": "built/index.d.ts",
   "scripts": {
-    "build": "npm run clean && tsc && tslint --project tsconfig.json --force",
+    "build": "tslint --project tsconfig.json && npm run clean && tsc",
```
Comment (tonyanziano) [inline packages/sdk/shared/package.json]: could be simplified to

`"npm run lint && npm run clean && tsc"`

File: packages/sdk/ui-react/src/styles/globalCss.ts [hunk 13]
```diff
@@ -32,20 +32,42 @@
...
+    css.global('::-webkit-scrollbar-track', {
+      background: Colors.SCROLLBAR_TRACK_BACKGROUND_DARK,
+    });
+    
+    css.global('::-webkit-scrollbar-thumb', {
+      background: Colors.SCROLLBAR_THUMB_BACKGROUND_DARK,
+    });
   }
-});
+}
```
Comment (tonyanziano) [inline packages/sdk/ui-react/src/styles/globalCss.ts]: missing newline

File: packages/extensions/qnamaker/client/src/App.tsx [hunk 14]
```diff
@@ -53,41 +53,15 @@ const AccessoryWorkingState = 'working';
...
-});
-
-css.global('::-webkit-scrollbar-track', {
-  background: Colors.SCROLLBAR_TRACK_BACKGROUND_DARK,
-});
-
-css.global('::-webkit-scrollbar-thumb', {
-  background: Colors.SCROLLBAR_THUMB_BACKGROUND_DARK,
-});
+GlobalCss.setCss(globalCss);
```
Comment (tonyanziano) [inline packages/extensions/qnamaker/client/src/App.tsx]: Seems to be breaking the build. There is no imported `GlobalCss`

Comment (justinwilaby) [reply packages/extensions/qnamaker/client/src/App.tsx]: +1

**Pull request merged**